### PR TITLE
PagedMemory: Use dicts to store items when there are only a few.

### PR DESCRIPTION
**Pull request by ltfish**
TODO:

- [x] Fix the type annotation for `DynamicDictList` to support `DynamicDictList[element_type]`.
- [x] Write a performance test case to determine the optimal threshold.

### CODE REVIEW

File: angr/storage/memory_mixins/paged_memory/pages/mv_list_page.py [hunk 1]
```diff
@@ -26,19 +27,21 @@ class MVListPage(
...
+        self.content: DynamicDictList[Optional[Union[_MOTYPE,Set[_MOTYPE]]]] = \
+            DynamicDictList(max_size=memory.page_size, content=content) if content is not None else None
         self.stored_offset = set()
         self._mo_cmp: Optional[Callable] = mo_cmp
 
-        if content is None:
+        if self.content is None:
             if memory is not None:
-                self.content: List[Optional[Union[_MOTYPE,Set[_MOTYPE]]]] = [None] * memory.page_size
+                self.content: DynamicDictList[Optional[Union[_MOTYPE,Set[_MOTYPE]]]] = \
```
Comment (rhelmot) [inline angr/storage/memory_mixins/paged_memory/pages/mv_list_page.py:37]: How does this type annotation work? What did you do to get the class to have type parameters like this?

**Review by rhelmot — APPROVED** [review]

File: angr/storage/memory_mixins/paged_memory/pages/mv_list_page.py (hunk 1 above)
Comment (rhelmot) [reply angr/storage/memory_mixins/paged_memory/pages/mv_list_page.py:37]: I see. this is one of the TODOs. never mind!

**Comment by ltfish** [comment]
Apparently this is never merged (or finished). I'll get to it today.

**Comment by ltfish** [comment]
My profiling script:

```python
import matplotlib.pyplot as plt
import timeit


NUMBER = 10000


def profile_list_dict(elements: int):
    t0 = timeit.Timer("dict(d)", setup=f"""
class A:
    pass

d = dict((v, A()) for v in range({elements}))""")

    t1 = timeit.Timer("list(l)", setup=f"""
class A:
    pass

l = [None] * 4096
for i in range({elements}):
    l[i] = A()""")

    d = t0.timeit(number=NUMBER)
    l = t1.timeit(number=NUMBER)
    print("#:", elements)
    print("dict:", d)
    print("list:", l)
    return d, l


def main():
    d_ = [ ]
    l_ = [ ]
    tt = [ ]
    for elements in range(0, 4096, 50):
        d, l = profile_list_dict(elements)
        tt.append(elements)
        d_.append(d)
        l_.append(l)

    # plot it
    fig, ax = plt.subplots()
    ax.plot(tt, d_, label="dict")
    ax.plot(tt, l_, label="list")
    ax.legend()

    plt.savefig("result.png")
    plt.show()


if __name__ == "__main__":
    main()
```

CPython 3.10:

cpy310

PyPy 3.9:

pypy39

Looks like we need two different thresholds :/

File: angr/storage/memory_mixins/paged_memory/pages/mv_list_page.py (hunk 1 above)
Comment (ltfish) [reply angr/storage/memory_mixins/paged_memory/pages/mv_list_page.py:37]: Fixed!

**Pull request merged**